### PR TITLE
dkg manager state transition

### PR DESCRIPTION
**Pull request by zjma**

### CODE REVIEW

File: dkg/src/network.rs [hunk 1]
```diff
@@ -181,7 +182,20 @@ impl RpcResponseSender for RealRpcResponseSender {
     }
 }
 
-#[cfg(test)]
```
Comment (zjma) [inline dkg/src/network.rs:184]: add it back

File: dkg/src/dkg_manager/mod.rs [hunk 2]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
+        mut self,
+        in_progress_session: Option<DKGSessionState>,
+        mut dkg_start_event_rx: aptos_channel::Receiver<(), DKGStartEvent>,
+        mut rpc_msg_rx: aptos_channel::Receiver<(), (AccountAddress, IncomingRpcRequest)>,
+        mut dkg_txn_pulled_rx: vtxn_pool::PullNotificationReceiver,
         close_rx: oneshot::Receiver<oneshot::Sender<()>>,
     ) {
+        let (agg_trx_tx, mut agg_trx_rx) = aptos_channel::new(QueueStyle::KLAST, 1, None);
+        self.agg_trx_tx = Some(agg_trx_tx);
+
```
Comment (zjma) [inline dkg/src/dkg_manager/mod.rs:113]: missing a `setup_deal_broadcast` call that consumes `in_progress_session` to handle crash recovery

File: dkg/src/dkg_manager/mod.rs [hunk 3]
```diff
@@ -1,30 +1,85 @@
...
 use std::sync::Arc;
 
-pub mod agg_trx_producer;
+#[derive(Clone, Debug)]
+enum InnerState<DKG: DKGTrait> {
+    NotStarted,
+    InProgress {
+        start_time_us: u64,
+        public_params: DKG::PublicParams,
+        my_node: DKGNode,
```
Comment (danielxiangzl) [inline dkg/src/dkg_manager/mod.rs:29]: Maybe add a comment saying the partially aggregated trx is in the broadcast status? It is a bit confusing at first glance that InProgress state only has local trx.

**Review by danielxiangzl — APPROVED** [review]

File: dkg/src/agg_trx_producer.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        &self,
+        epoch_state: Arc<EpochState>,
+        dkg_config: S::PublicParams,
+        agg_trx_tx: Option<Sender<(), S::Transcript>>,
+    ) -> AbortHandle;
+}
+
+/// The real implementation of `AggTranscriptProducer` that broadcasts a `NodeRequest`, collects and verifies nodes from network.
+pub struct RealAggTranscriptProducer {
+    reliable_broadcast: Arc<ReliableBroadcast<DKGMessage, ExponentialBackoff>>,
```
Comment (danielxiangzl) [inline dkg/src/agg_trx_producer.rs:29]: We may need to revisit the `ExponentialBackoff` here. Nodes may be less synchronized about the DKG start time and computation time.

File: dkg/src/agg_trx_producer.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    fn start_produce(
+        &self,
+        epoch_state: Arc<EpochState>,
+        dkg_config: S::PublicParams,
+        agg_trx_tx: Option<Sender<(), S::Transcript>>,
+    ) -> AbortHandle;
+}
+
+/// The real implementation of `AggTranscriptProducer` that broadcasts a `NodeRequest`, collects and verifies nodes from network.
+pub struct RealAggTranscriptProducer {
```
Comment (zekun000) [inline dkg/src/agg_trx_producer.rs:28]: I probably started the bad name convention 🤦 typically we name the trait TAggTranscriptProducer, and the struct AggTranscriptProducer

File: dkg/src/agg_trx_producer.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+/// A sub-process of the whole DKG process.
+/// Once invoked by `DKGManager` to `start_produce`,
+/// it starts producing an aggregated transcript and returns an abort handle.
+/// Once an aggregated transcript is available, it is sent back via channel `agg_trx_tx`.
+pub trait AggTranscriptProducer<S: DKGTrait>: Send + Sync {
+    fn start_produce(
+        &self,
+        epoch_state: Arc<EpochState>,
+        dkg_config: S::PublicParams,
+        agg_trx_tx: Option<Sender<(), S::Transcript>>,
```
Comment (zekun000) [inline dkg/src/agg_trx_producer.rs:23]: why this is Option?

File: dkg/src/dkg_manager/mod.rs [hunk 7]
```diff
@@ -1,30 +1,85 @@
...
 use futures_channel::oneshot;
-use futures_util::{FutureExt, StreamExt};
+use futures_util::{future::AbortHandle, FutureExt, StreamExt};
 use move_core_types::account_address::AccountAddress;
+use rand::thread_rng;
 use std::sync::Arc;
 
-pub mod agg_trx_producer;
+#[derive(Clone, Debug)]
+enum InnerState<DKG: DKGTrait> {
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:24]: why it's not just DKGState?

File: dkg/src/dkg_manager/mod.rs [hunk 8]
```diff
@@ -1,30 +1,85 @@
...
+
+impl<DKG: DKGTrait> Default for InnerState<DKG> {
+    fn default() -> Self {
+        Self::NotStarted
+    }
+}
 
 #[allow(dead_code)]
 pub struct DKGManager<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> {
     private_params_provider: P,
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:47]: does this need to be a separate trait? or just DKG::PrivateParams

File: dkg/src/agg_trx_producer.rs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+use futures::future::AbortHandle;
+use futures_util::future::Abortable;
+use std::sync::Arc;
+use tokio_retry::strategy::ExponentialBackoff;
+
+/// A sub-process of the whole DKG process.
+/// Once invoked by `DKGManager` to `start_produce`,
+/// it starts producing an aggregated transcript and returns an abort handle.
+/// Once an aggregated transcript is available, it is sent back via channel `agg_trx_tx`.
+pub trait AggTranscriptProducer<S: DKGTrait>: Send + Sync {
```
Comment (zekun000) [inline dkg/src/agg_trx_producer.rs:18]: why this needs to be trait? for unit test?

File: dkg/src/dkg_manager/mod.rs [hunk 10]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
     }
 
     pub async fn run(
-        self,
-        _in_progress_session: Option<DKGSessionState>,
-        _dkg_start_event_rx: aptos_channel::Receiver<(), DKGStartEvent>,
-        _rpc_msg_rx: aptos_channel::Receiver<(), (AccountAddress, IncomingRpcRequest)>,
-        _dkg_txn_pulled_rx: vtxn_pool::PullNotificationReceiver,
+        mut self,
+        in_progress_session: Option<DKGSessionState>,
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:105]: I don't see this used? is it for crash recovery? can't we recover from re-sending the DKGStartEvent?

File: dkg/src/dkg_manager/mod.rs [hunk 11]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
+        mut dkg_start_event_rx: aptos_channel::Receiver<(), DKGStartEvent>,
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:106]: can this be a one-shot channel since it's once per epoch?

File: dkg/src/dkg_manager/mod.rs [hunk 12]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
+        target_validator_set: &ValidatorSet,
+    ) -> Result<()> {
+        self.state = match &self.state {
+            InnerState::NotStarted => {
+                let public_params = DKG::new_public_params(
+                    self.epoch_state.as_ref(),
+                    self.my_addr,
+                    target_validator_set,
+                );
+                let mut rng = thread_rng();
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:191]: I think I asked the question before but does this need to be persisted? or different values matters? e.g. crash recovery produces different transcript?

File: dkg/src/dkg_manager/mod.rs [hunk 13]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
+                let trx = DKG::generate_transcript(
+                    &mut rng,
+                    self.private_params_provider.dkg_private_params(),
+                    &public_params,
+                );
+
+                let dkg_node = DKGNode::new(
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:198]: DKGNode is a bad name 🤦 just call it DKGTranscript?

File: dkg/src/dkg_manager/mod.rs [hunk 14]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
+            InnerState::InProgress {
+                start_time_us,
+                my_node,
+                ..
+            } => {
+                // TODO(zjma): metric DKG_AGG_NODE_READY
+                let txn = ValidatorTransaction::DKGResult(DKGNode {
+                    metadata: DKGTranscriptMetadata {
+                        epoch: self.epoch_state.epoch,
+                        author: self.my_addr,
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:238]: it looks different node will produce different DKG transaction? we don't deduplicate from them?

File: dkg/src/epoch_manager.rs [hunk 15]
```diff
@@ -89,24 +90,42 @@ impl<P: OnChainConfigProvider> EpochManager<P> {
...
-        _peer_id: AccountAddress,
-        _dkg_request: IncomingRpcRequest,
+        peer_id: AccountAddress,
+        dkg_request: IncomingRpcRequest,
     ) -> Result<()> {
-        //TODO
+        if Some(dkg_request.msg.epoch()) == self.epoch_state.as_ref().map(|s| s.epoch) {
+            // Forward to DKGManager if it is alive.
+            if let Some(tx) = &self.dkg_rpc_msg_tx {
+                tx.push((), (peer_id, dkg_request));
```
Comment (zekun000) [inline dkg/src/epoch_manager.rs:99]: why the key is ()? we use the peer_id for DDoS prevention

File: dkg/src/dkg_manager/mod.rs [hunk 16]
```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
...
+                dkg_txn = dkg_txn_pulled_rx.select_next_some() => {
+                    self.process_dkg_txn_pulled_notification(dkg_txn).await
+                },
                 close_req = close_rx.select_next_some() => {
-                    self.vtxn_pool_write_cli.put(None);
-                    if let Ok(ack_sender) = close_req {
-                        ack_sender.send(()).unwrap();
-                    }
-                    break;
+                    self.process_close_cmd(close_req.unwrap())
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:130]: nit: I think we should just tolerate the close_req being Err (the sender was dropped) for sanity

File: dkg/src/dkg_manager/mod.rs (hunk 12 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs:191]: now nothing needs to be persisted. Any trx is fine as long as it's verifiable.

Comment (zekun000) [reply dkg/src/dkg_manager/mod.rs:191]: worth adding a comment here

File: dkg/src/dkg_manager/mod.rs (hunk 14 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs:238]: that's the cost of we saving a DB. It's hard to duplicate now.

File: dkg/src/agg_trx_producer.rs (hunk 9 above)
Comment (zjma) [reply dkg/src/agg_trx_producer.rs:18]: yes

File: dkg/src/dkg_manager/mod.rs (hunk 10 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs:105]: yes it's for crash recovery. I thought I pushed a fix.
How to re-send DKGStart?

File: dkg/src/dkg_manager/mod.rs (hunk 12 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs:208]: yes, we have some comments not far, on top of `setup_deal_broadcast`

File: dkg/src/dkg_manager/mod.rs (hunk 13 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs]: fixed

File: dkg/src/epoch_manager.rs (hunk 15 above)
Comment (zjma) [reply dkg/src/epoch_manager.rs]: fixed

File: dkg/src/dkg_manager/mod.rs (hunk 16 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs]: fixed

File: dkg/src/dkg_manager/mod.rs [hunk 17]
```diff
@@ -122,11 +130,12 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
         let (agg_trx_tx, mut agg_trx_rx) = aptos_channel::new(QueueStyle::KLAST, 1, None);
         self.agg_trx_tx = Some(agg_trx_tx);
 
+        let mut dkg_start_event_rx = dkg_start_event_rx.into_stream();
         let mut close_rx = close_rx.into_stream();
         while !self.stopped {
             let handling_result = tokio::select! {
                 dkg_start_event = dkg_start_event_rx.select_next_some() => {
-                    self.process_dkg_start_event(dkg_start_event).await
+                    self.process_dkg_start_event(dkg_start_event.ok()).await
```
Comment (zekun000) [inline dkg/src/dkg_manager/mod.rs:138]: why this is Option?

**Review by zekun000 — APPROVED** [review]

File: dkg/src/dkg_manager/mod.rs (hunk 17 above)
Comment (zjma) [reply dkg/src/dkg_manager/mod.rs:138]: since now it's a oneshot, `dkg_start_event` is a result?

Comment (zekun000) [reply dkg/src/dkg_manager/mod.rs:138]: hmm, we should shutdown if it's error?

Comment (zjma) [reply dkg/src/dkg_manager/mod.rs:146]: it's error only when epoch manager is dropped or epoch manager dropped the tx, both should be unreachable?

**Pull request merged**